app/helpers/Evaluation/ScoreCalculators/UniversalScoreCalculator/AstNode.php

```diff
@@ -0,0 +1,196 @@
+     * This identifier is used in (de)serialization.
+     * @return string
+     */
+    public function getTypeName(): string
+    {
+        return static::$TYPE_NAME;
+    }
+
+    // Internal data common to all nodes
+    protected $parent = null; // reference to parent node
+    protected $children = []; // list of subnodes
+    protected $associatedData = []; // keeps extra data (with `x-` prefix) from deserialization, so they are not lost
+
+    /**
+     * Create and initialize empty instance of the node.
+     * @param array $config Optional parameter used in deserialization.
+     *                      Data from $config are deserialized into the node.
+     * @throws AstNodeException if config data are not valid and deserialization fails
+     */
+    public function __construct(array $config = [])
+    {
+        // load children recursively
+        if (array_key_exists(self::KEY_CHILDREN, $config)) {
+            if (!is_array($config[self::KEY_CHILDREN])) {
+                throw new AstNodeException("AST node children must be represented as array.");
+            }
+
+            foreach ($config[self::KEY_CHILDREN] as $childConfig) {
+                if (is_numeric($childConfig)) {
+                    // the only special case in the config (number literal may be encoded as scalar values)
+                    $child = new AstNodeValue();
+                    $child->setValue((float)$childConfig);
+                } else {
+                    // regular case
+                    if (!is_array($childConfig)) {
+                        throw new AstNodeException("AST node configuration must be an associative array.");
+                    }
+                    $child = self::createFromConfig($childConfig);
+                }
+                $this->addChild($child);
+            }
+        }
+
+        $this->associatedData = array_filter($config, function ($key) {
+            return substr($key, 0, 2) === 'x-'; // only extension fields are kept (without interpreting)
+        }, ARRAY_FILTER_USE_KEY);
+    }
+
+    /**
+     * Append another child node to this node.
+     * @param AstNode $child New node to be appended
+     */
+    public function addChild(AstNode $child)
+    {
+        if ($child->parent) {
+            throw new AstNodeException("Unable to add child to a node. The child already belongs to another node.");
+        }
+        $child->parent = $this;
+        $this->children[] = $child;
+    }
+
+    /**
+     * Perform internal validation of the node itself (not the subtree).
+     * This method stands aside, so it can be reused and the main interface
+     * (validate() method) does not have to be overriden in derived classes.
+     * @param array $testNames List of allowed test names (array of strings)
+     * @throws AstNodeException if something is not valid
+     */
+    protected function internalValidation(array $testNames = [])
+    {
+        // expected to be overriden in derived classes
```

Comment:
If it is expected to be overriden, why is it not abstract?

Reply:
Because it is expected but not enforced. Empty implementation is the desired default.